### PR TITLE
CLDR-13654 units: add consistency examples and checks, and ability to generate compound names to that end.

### PR DESCRIPTION
**Pull request by macchiati**
##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/CLDR-13654
- [x] Updated PR title and link in previous line to include Issue number

### CODE REVIEW

**Comment by macchiati** [comment]
Adding generation of display names and gender for units. Goal is to then use these in examples.

**Comment by macchiati** [comment]
Actually, hold off on reviewing until I can describe the changes.

**Comment by macchiati** [comment]
Overview of changes for review.

Note that this involved cleaning up some of the code; moving methods from one class to a more appropriate place.

- common/main/en.xml  fix bug in English : inconsistency between pattern and long explicit compound unit names.
- TestExampleGenerator.java : add test of new check on consistency between unit components (kilo- ... square-...) and long explicit compound units.
- tools/cldr-unittest/src/org/unicode/cldr/unittest/TestUnits.java Significant restructuring to add TestGender(), TestFallbackNames(), TestCheckUnits()
- tools/java/org/unicode/cldr/test/CheckCLDR.java added more check enums
- tools/java/org/unicode/cldr/test/CheckUnits.java - cleanup of code using XPathParts and UnitPathType. Added check for power and prefixes consistency (generates warnings, not errors)
- tools/java/org/unicode/cldr/test/ExampleGenerator.java - added more examples for power and prefix. OverridingStringProvider is not yet in use, but planned for future cases.
- tools/java/org/unicode/cldr/util/CLDRFile.java Added interface to allow easy overrides for ExampleGenerator. Move getGrammaticalInfoAttributes to GrammarInfo.
- tools/java/org/unicode/cldr/util/GrammarInfo.java - substantial revisions
- tools/java/org/unicode/cldr/util/LocaleStringProvider.java interface for CLDRFile.
- tools/java/org/unicode/cldr/util/UnitConverter.java - fair amount of restructuring, focused on more correctly generating compound unit names, particularly in UnitId's toString (line 725). Also needed to add computation of gender from components. Later I foresee adding a data structure to control this, when different languages have different algorithms.
- tools/java/org/unicode/cldr/util/UnitPathType.java - moved / developed a lot of code here, since it allows us to centralize a lot of logic around the different kinds of paths used in unit processing. I also hang examples off of the enums for use in the example generator and other places (again for centralization).

**Comment by macchiati** [comment]
tests fail; must have messed something up in resolving the merge issues. Won't be able to fix right away.

**Comment by macchiati** [comment]
More cleanup.  Mostly resolution of merge issues.  See Overview of changes for review above, plus:

tools/cldr-unittest/src/org/unicode/cldr/unittest/TestCLDRFile.java
A couple of small changes for debugging problems.

tools/cldr-unittest/src/org/unicode/cldr/unittest/TestPathHeader.java
A couple of small changes for debugging problems.

tools/java/org/unicode/cldr/util/CLDRFile.java
Moved in OverridingStringProvider and enhanced

tools/java/org/unicode/cldr/util/UnitPathType.java
Pulled in a couple of core features that were statics elsewhere

File: tools/java/org/unicode/cldr/util/UnitConverter.java [hunk 1]
```diff
@@ -646,6 +727,175 @@ public String toString() {
...
+                    if (deprefix.value != 1) {
+                        prefixPattern = UnitPathType.prefix.getTrans(resolvedFile, width, "10p" + deprefix.value, _pluralCategory, caseVariant, genderVariant, partsUsed);
+                    }
+
+                    // get the core pattern. Detect and remove the the placeholder (and surrounding spaces)
+                    String unitPattern = UnitPathType.unit.getTrans(resolvedFile, width, unit, pluralCategory, caseVariant, genderVariant, partsUsed);
+                    if (unitPattern == null) {
+                        return null; // unavailable
+                    }
+                    // we are set up for 2 kinds of placeholder patterns for units. {0}\s?stuff or stuff\s?{0}, or nothing(Eg Arabic)
```
Comment (pedberg-icu) [inline tools/java/org/unicode/cldr/util/UnitConverter.java:801]: There are units that do not follow those 3 types, for example in Korean:
```
    <unit type="concentr-milligram-per-deciliter">
        <displayName>데시리터당 밀리그램</displayName>
        <unitPattern count="other">데시리터당 {0}밀리그램</unitPattern>
    </unit>
```

**Comment by macchiati** [comment]
You're right. However, the code will only be called on simpler units (no
times or per), and none of those simpler units have infix placeholders.

I should document that, but would you mind if I did that on a follow-up PR?

**Comment by pedberg-icu** [comment]
> You're right. However, the code will only be called on simpler units (no times or per), and none of those simpler units have infix placeholders. I should document that, but would you mind if I did that on a follow-up PR?

Fine to document in another PR, I just wanted to make sure this code would not get called for that type of pattern.

**Pull request merged**